### ISSUE TITLE
What is the reason for remove the class token during the unshuffling and then append it back in the decoder?

### ISSUE THREAD

**Issue by scaomath** [facebookresearch/mae#23]
https://github.com/facebookresearch/mae/blob/6a2ba402291005b003a70e99f7c87d1a2c376b0d/models_mae.py#L178-L180

Right here in lines above.

**Comment by yaohuicai** [comment]
This is because the `cls` token is appended to the token sequence after doing `random_maksing`. So the `ids_restore` does not include the id for `cls` token.

https://github.com/facebookresearch/mae/blob/6a2ba402291005b003a70e99f7c87d1a2c376b0d/models_mae.py#L158-163

Thus when inserting the `mask_token` (the code you showed), you want to drop the `cls` token first.

**Comment by KaimingHe** [comment]
It is not a must, and there are many other alternative ways. The main motivation here is to avoid many corners cases when the cls token is masked, for example, when visualizing the mask, when padding mask tokens in the decoder, and when computing the loss on the unknown tokens. All these corners can be relatively easily handled, but the code would become more cumbersome and would become less readable as it would be dominated by handling this cls token.

The cls token can also be completely removed from pre-training and fine-tuning.
